tools/flow_grid.py

```diff
@@ -0,0 +1,689 @@
+        'se' : {'k' : tuple(indices[:,-1,-1]),
+                'v' : [[-1,-2,-2], [-2,-2,-1]],
+                'pad': np.array([7,8,1])}
+        }
+    
+        # initialize indices of edges
+        edge = {
+        'n' : {'k' : tuple(indices[:,0,1:-1]),
+               'pad' : np.array([3,4,5,6,7])},
+        'w' : {'k' : tuple(indices[:,1:-1,0]),
+               'pad' : np.array([1,2,3,4,5])},
+        'e' : {'k' : tuple(indices[:,1:-1,-1]),
+               'pad' : np.array([1,5,6,7,8])},
+        's' : {'k' : tuple(indices[:,-1,1:-1]),
+               'pad' : np.array([1,2,3,7,8])}
+        }
+    
+        # initialize indices of body (all cells except edges and corners)
+        body = indices[:, 1:-1, 1:-1]
+    
+        # initialize output array
+        outmap = np.full(self.shape, nodata, dtype=np.int8)
+    
+        # select the eight cells surrounding a cell
+        def select_surround(i, j):
+            return ([i-1, i-1, i+0, i+1, i+1, i+1, i+0, i-1],
+                   [j+0, j+1, j+1, j+1, j+0, j-1, j-1, j-1])
+    
+        # select the five cells surrounding an edge cell
+        def select_edge_sur(k):
+            i,j = edge[k]['k']
+            if k == 'n':
+                return [i+0, i+1, i+1, i+1, i+0], [j+1, j+1, j+0, j-1, j-1]
+            elif k =='e':
+                return [i-1, i+1, i+1, i+0, i-1], [j+0, j+0, j-1, j-1, j-1]
+            elif k =='s':
+                return [i-1, i-1, i+0, i+0, i-1], [j+0, j+1, j+1, j-1, j-1]
+            elif k == 'w':
+                return [i-1, i-1, i+0, i+1, i+1], [j+0, j+1, j+1, j+1, j+0]
+     
+        # for each entry in "body" determine flow direction based
+        # on steepest neighboring slope
+        for i, j in np.nditer(tuple(body), flags=['external_loop']):
+            dat = data[i,j]
+            sur = data[select_surround(i,j)]
+            a = ((dat - sur) > 0).any(axis=0)
+            b = np.argmax((dat - sur), axis=0) + 1
+            c = flat
+            outmap[i,j] = np.where(a,b,c)
+
+        # determine flow direction for edges and corners, if desired
+        if include_edges == True:
+
+            # fill corners
+            for i in corner.keys():
+                dat = data[corner[i]['k']]
+                sur = data[corner[i]['v']]
+                if ((dat - sur) > 0).any():
+                    outmap[corner[i]['k']] = corner[i]['pad'][np.argmax(dat - sur)]
+                else:
+                    outmap[corner[i]['k']] = flat
+
+            # fill edges
+            for x in edge.keys():
+                dat = data[edge[x]['k']]
+                sur = data[select_edge_sur(x)]
+                a = ((dat - sur) > 0).any(axis=0)
+                b = edge[x]['pad'][np.argmax((dat - sur), axis=0)]
+                c = flat
+                outmap[edge[x]['k']] = np.where(a,b,c)
+    
+        # If direction numbering isn't default, convert values of output array.
+        if dirmap != (1,2,3,4,5,6,7,8):
```

Comment:
maybe consider a two module level global variables for the dirmap:
```python
dirmap_1_8 = (1,2,3,4,5,6,7,8)
dirmap_1_128 = (1,2,4,8,16,32,64,128)
```
then replace the hard-coded tuples you have throughout the module.